hydra-cluster/src/CardanoNode.hs

```diff
@@ -125,6 +125,28 @@ getCardanoNodeVersion :: IO String
 getCardanoNodeVersion =
   readProcess "cardano-node" ["--version"] ""
 
+-- | Tries to reach for an existing cardano-node running on default state directory.
+findRunningCardanoNode :: FilePath -> KnownNetwork -> IO (Maybe RunningNode)
+findRunningCardanoNode stateDirectory knownNetwork = do
+  let args'@CardanoNodeArgs{nodeSocket} = defaultCardanoNodeArgs
+  let nodeSocketPath = File nodeSocket
+  maybeSocketPath <- doesFileExist $ unFile nodeSocketPath
+  if maybeSocketPath
+    then do
+      let
+        knownNetworkId = case knownNetwork of
+          Mainnet -> Api.Mainnet
+          Preproduction -> Api.Testnet (Api.NetworkMagic 1)
+          Preview -> Api.Testnet (Api.NetworkMagic 2)
+          Sanchonet -> Api.Testnet (Api.NetworkMagic 4)
+      try (queryTip knownNetworkId nodeSocketPath)
+        >>= \case
+          Right _ -> pure Nothing
+          Left (_ :: SomeException) -> do
+            n@RunningNode{networkId} <- mkRunningNode stateDirectory nodeSocketPath args'
+            pure $ if knownNetworkId == networkId then Just n else Nothing
```

Comment:
Sould: use `guard`

Reply:
Ah.. it's `IO (Maybe RunningNode)`. So it would be `pure $ guard (knownNetworkId == networkId) >> pure n`

But this is not much clearer indeed.